pkg/asset/manifests/openshift.go

```diff
@@ -69,6 +69,9 @@ func (o *Openshift) Dependencies() []asset.Asset {
 		&openshift.PrivateClusterOutbound{},
 		&openshift.BaremetalConfig{},
 		new(rhcos.Image),
+		&openshift.AzureCloudProviderSecret{},
```

Comment:
@patrickdillon an issue with the appraoch of using templates is this: we have to declare templates as dependencies and store will call `Generate()` on them even if we don't need them (e.g. non-ARO builds).

What do you think? Is it something we should worry about?

We can probably use build tags and create two extra files in this package:

1. One with `// +build aro` where we include them as dependencies
2. One with `// +build !aro` where we do not include them

But I don't want to rely on build tags too much in this case:

1. It will make the code harder to follow
2. It will probably be unnecessary once you implement Azure Stack Hub support

Reply:
If we're going with the template approach, I was thinking to just add them to the existing templates. Here's a simplified example: https://github.com/openshift/installer/compare/master...patrickdillon:azure-merge-cloud-config We would need to add a boolean field to `AzureCredsSecretData` to switch this behavior on or off depending on ASH/ARO. 

Also, not sure if it is valid to have the values of the cloud provider config secret be base64 encoded. I assume for the existing code, the cloud credential operator knows to base64 decode the values, but I assume that is up to the client. I don't see anything in the azure docs about this. If we don't know, we can test and see.

I would like to get @staebler's feedback about whether this is the right direction with templates. To summarize the context: In the case of ARO & ASH we need to create a secret (with the creds) , role, and role binding. I see a few options:

1. Use the template approach, like in my [example above](https://github.com/openshift/installer/compare/master...patrickdillon:azure-merge-cloud-config )
2. Generate all 3 resources in the cloud provider config, which was the original state of this PR
3. Some combination of the above
4. Generate the resource manifests somewhere else entirely

I'm pretty neutral on these options with a slight leaning toward option 1.

Reply:
@patrickdillon thanks for the feedback. Replied to some of the comments.

> If we're going with the template approach, I was thinking to just add them to the existing templates. Here's a simplified example: [master...patrickdillon:azure-merge-cloud-config](https://github.com/openshift/installer/compare/master...patrickdillon:azure-merge-cloud-config) We would need to add a boolean field to `AzureCredsSecretData` to switch this behavior on or off depending on ASH/ARO.

I was thinking about reusing `data/data/manifests/openshift/cloud-creds-secret.yaml.template`, but it is already overloaded with logic and I didn't want to make it worse. Happy to do this if you think it is a better appraoch.

> Also, not sure if it is valid to have the values of the cloud provider config secret be base64 encoded. I assume for the existing code, the cloud credential operator knows to base64 decode the values, but I assume that is up to the client. I don't see anything in the azure docs about this. If we don't know, we can test and see.

We must encode values into base64 manually here because:
* We add them into [`Secret`](https://kubernetes.io/docs/concepts/configuration/secret/#overview-of-secrets)'s `data`:
  > The values for all keys in the `data` field have to be base64-encoded strings.
* We use templates. Normaly when marshaling into json/yaml from a go struct it will be done implicitly due to the fact that `data` is defined as follows:

  ```go
  Data map[string][]byte `json:"data,omitempty" protobuf:"bytes,2,rep,name=data"`
  ```

  And [`[]byte` in go marshaled](https://golang.org/pkg/encoding/json/#Marshal) as base64-encoded string.

  > Array and slice values encode as JSON arrays, except that []byte encodes as a base64-encoded string, and a nil slice encodes as the null JSON value. 


On the receiving end, the cloud provider doesn't have to decode the value explicilty [here](https://github.com/openshift/origin/blob/90c050f5afb4c52ace82b15e126efe98fa798d88/vendor/k8s.io/legacy-cloud-providers/azure/azure_config.go#L77): go will do it behind the scenes.

Reply:
> I was thinking about reusing data/data/manifests/openshift/cloud-creds-secret.yaml.template, but it is already overloaded with logic and I didn't want to make it worse. Happy to do this if you think it is a better appraoch.

As I look at this more closely I do think this current approach is on the right track. I think the approach should very closely follow the pattern here: 
https://github.com/openshift/installer/blob/master/pkg/asset/manifests/openshift.go#L253-L259

> we have to declare templates as dependencies and store will call Generate() on them even if we don't need them (e.g. non-ARO builds).

So yes this is fine to have a dependency generated even when it is not used. The above example is only for private clusters, but is generated for all clusters. But you should consolidate all of these Azure Cloud Provider dependencies into a single asset. Note that the one asset could have many templates; so you could either have a separate template for each resource or combine the three resources in a single template. 

If this is unclear or your have questions let me know. I think this is the right drection!

Also, thanks for the detailed, awesome explanation of why secret values are b64 encoded. I didn't know that!

Reply:
@patrickdillon I left 3 template files, but consolidated them into one assset. Now the naming is little bit tricky: I called it `AzureCloudProviderSecret`, but it contains the role, the role binding as well as the secret which might be a little bit confusing. Suggestions are welcome :)

I'm also happy to put all yamls into one file, if you like. Having in separate templates/manifests is a tiny bit better for visibility, IMO. But I have no strong prefrences.

Reply:
I think this is looking good. Have you tested this? I hope to test it soon with Azure Stack.

> I called it AzureCloudProviderSecret... Suggestions are welcome :)

Maybe `AzureMergedCloudConfig`?

Reply:
@patrickdillon

> Have you tested this?

I did `openshift-install create manifests` with `aro` build tag and manifests look good. I alos did it with a binary compiled without `aro` build tag and relevant manifests were not present in the output directory, as expected.

And I'm pretty confident that it will work with ARO because I tested it with older version of this PR (we did not change outputs significantly in recent changes). I'll ARO integration tomorrow, but it should not block this PR.


> Maybe `AzureMergedCloudConfig`?

I was wonder if we can better reflect the fact that it is not only secret, but role and role binding as well.
Let's probably keep it as `AzureCloudProviderSecret` as it consists of resource name `azure-cloud-provider` and resource type `secret`. Resource name is [forced on us by cloud provider](https://kubernetes-sigs.github.io/cloud-provider-azure/install/configs/#setting-azure-cloud-provider-from-kubernetes-secrets).

Reply:
> I'm also happy to put all yamls into one file, if you like. Having in separate templates/manifests is a tiny bit better for visibility, IMO. But I have no strong prefrences.

Leave it as one file for each resource. The CVO has (or had) issues with multiple resources in a single manifest when there are problems applying some of the resources.

Reply:
> We must encode values into base64 manually here because:
>
>    We add them into Secret's data:

Use `stringData` instead of `data`. Then you don't need to do the base-64 encoding. The API server will do the base-64 encoding for you.

Reply:
I didn't know that `stringData` merges into `data`.

@staebler I updated the PR to use `stringData`. Note that everything else in `pkg/asset/manifests/openshift.go` explicitly encodes into base64. I'm happy to revert back to explicit encoding for consistency, if you wish.